### PR TITLE
C++ user-defined functions (UDF & UDAF) (not ready for production!!!)

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Adding three functions `udf`, `udsf`, and `udaf` to support user-defined C++ functions in SQL.

**This PR is to demonstrate & discuss the idea. It is not ready for merging / production use.**

### CODE REVIEW

File: dbms/src/Interpreters/UDFHeaders.h [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+    #_0 ", " #_1 ", " #_2 ", " #_3 ", " #_4 ", " #_5 "\n"
+#define CODE_TO_STR_IMPL_7(_0, _1, _2, _3, _4, _5, _6) \
+    #_0 ", " #_1 ", " #_2 ", " #_3 ", " #_4 ", " #_5 ", " #_6 "\n"
+#define CODE_TO_STR_IMPL_8(_0, _1, _2, _3, _4, _5, _6, _7) \
+    #_0 ", " #_1 ", " #_2 ", " #_3 ", " #_4 ", " #_5 ", " #_6 ", " #_7 "\n"
+
+#define CODE_TO_STR_SELECT(_0, _1, _2, _3, _4, _5, _6, _7, MACRO, ...) \
+    MACRO
+
+#define CODE_TO_STR(...) \
```
Comment (amosbird) [inline dbms/src/Interpreters/UDFHeaders.h]: using raw string literals instead?

**Comment by amosbird** [comment]
Interesting interface. At first glance I was thinking that we can have dynamic UDFs storing the code strings along with the data. It seems not yet possible though. Since we are already doing code generation by text manipulation, shall we use sized array instead of std::vector for column data?

**Comment by alexey-milovidov** [comment]
I think that allowing the user to write C++ code in SQL statements is not viable.
Better to only allow reading the code from files in configured location in filesystem.

We should not put all functions inside "udf" name. They should be registered with user defined names by `CREATE FUNCTION` query and appear in `system.functions` table.

Isolation of processes must be implemented. For example, IColumn may map its data to shared memory. Please also consider alternative approaches: https://issues.apache.org/jira/browse/SPARK-21190 https://issues.apache.org/jira/browse/SPARK-21404

Please consider how the user will be able to use third-party libraries in their functions.

**Comment by filimonov** [comment]
apache arrow gives and interesting toolset to build that:
https://www.slideshare.net/ueshin/apache-arrow-and-pandas-udf-on-apache-spark

**Comment by hczhcz** [comment]
> shall we use sized array instead of std::vector for column data?

The data is passed per-block. If the array size is given in compile-time, it will require re-compiling for different blocks.
As an alternative, we may pass the array size and a pointer to the data. The performance will be much better than std::vector but it will add some difficulty to UDF code writing. Actually, with the help of C++'s type system, it is possible to support both.

```c++
void f(size_t row_size, int * result, int * a, int * b);
// vs.
vector<int> f(vector<int> a, vector<int> b);
```

> They should be registered with user defined names by CREATE FUNCTION query and appear in system.functions table.

I think so. This PR is just a conceptual thing.

Some possible syntax to consider:

```sql
create function f -- and also `create aggregate function`, `create table function`, etc?
language C++ -- maybe multiple language engines such as "C++PerRow" vs "C++" ("C++PerBlock"), or even "C++Raw" for direct pointer access?
code ' // maybe some better string literal syntax?
    int f(int a, int b) {
        return a + b;
    }
'

create function f
language C++ codefile 'file_from_given_location.cpp'
```

**Comment by alexey-milovidov** [comment]
I thought about something like

```
CREATE FUNCTION f ... ENGINE = SQL
```

Where engine is mapped to something like `IUserDefinedFunctionCreator` the main responsibility is to register a new function in FunctionFactory.

PS. C++ UDFs should be the last thing that we implement.
The main reason is debug-ability of 3rd party ClickHouse installations.

**Comment by alex-krash** [comment]
A stupid question, may the following solution exist:
1.) Compiling UDF into a separate .so, using header with UDF API from CH source
2.) When executing, perform fork (or use pre-forked pool of instances), and load .so
3.) Setup communication between CH server and UDF executor via pipes/MQ protocol/files/sockets
4.) Execute UDF on executor side and give results back to CH server.

**Comment by alex-krash** [comment]
Or, even more crazy idea - "ExternalUDFStream". Write incoming data into RowBinary format into socket/pipe/mmaped files/MQ/stdin for "worker", that can be implemented in any programming language.

**Comment by alexey-milovidov** [comment]
> Or, even more crazy idea - "ExternalUDFStream". Write incoming data into RowBinary format into socket/pipe/mmaped files/MQ/stdin for "worker", that can be implemented in any programming language.

It is one of the working ideas. Probably the most safe but least performant solution. BTW, it's already available via executable dictionaries with cache layout but it doesn't even reach the potential of this implementation way.

> When executing, perform fork (or use pre-forked pool of instances), and load .so

Better to avoid fork of running process due to storm of page fault. vfork + exec of a fresh process is Ok.

**Comment by compasses** [comment]
Since it was closed, CH support UDF now? Or any docs about it? I believe it will be a great win if CH support extension, which user can define, mostly the high level user.
Recently i want to implement some udf, and found i have lots of things to do.

**Pull request closed without merging**